cli/ping.go

```diff
@@ -77,11 +143,62 @@ func (r *RootCmd) ping() *serpent.Command {
 			defer conn.Close()
 
 			derpMap := conn.DERPMap()
-			_ = derpMap
+
+			diagCtx, diagCancel := context.WithTimeout(inv.Context(), 30*time.Second)
+			defer diagCancel()
+			diags := conn.GetPeerDiagnostics()
+			cliui.PeerDiagnostics(inv.Stdout, diags)
+
+			// Silent ping to determine whether we should show diags
+			_, didP2p, _, _ := conn.Ping(ctx)
+
+			ni := conn.GetNetInfo()
+			connDiags := cliui.ConnDiags{
+				DisableDirect:      r.disableDirect,
+				LocalNetInfo:       ni,
+				Verbose:            r.verbose,
+				PingP2P:            didP2p,
+				TroubleshootingURL: appearanceConfig.DocsURL + "/networking/troubleshooting",
+			}
+
+			awsRanges, err := cliutil.FetchAWSIPRanges(diagCtx, cliutil.AWSIPRangesURL)
+			if err != nil {
+				opts.Logger.Debug(inv.Context(), "failed to retrieve AWS IP ranges", slog.Error(err))
+			}
+
+			connDiags.ClientIPIsAWS = isAWSIP(awsRanges, ni)
+
+			connInfo, err := wsClient.AgentConnectionInfoGeneric(diagCtx)
+			if err != nil || connInfo.DERPMap == nil {
+				return xerrors.Errorf("Failed to retrieve connection info from server: %w\n", err)
+			}
+			connDiags.ConnInfo = connInfo
+			ifReport, err := healthsdk.RunInterfacesReport()
+			if err == nil {
+				connDiags.LocalInterfaces = &ifReport
+			} else {
+				_, _ = fmt.Fprintf(inv.Stdout, "Failed to retrieve local interfaces report: %v\n", err)
+			}
+
+			agentNetcheck, err := conn.Netcheck(diagCtx)
+			if err == nil {
+				connDiags.AgentNetcheck = &agentNetcheck
+				connDiags.AgentIPIsAWS = isAWSIP(awsRanges, agentNetcheck.NetInfo)
+			} else {
+				var sdkErr *codersdk.Error
+				if errors.As(err, &sdkErr) && sdkErr.StatusCode() == http.StatusNotFound {
+					_, _ = fmt.Fprint(inv.Stdout, "Could not generate full connection report as the workspace agent is outdated\n")
+				} else {
+					_, _ = fmt.Fprintf(inv.Stdout, "Failed to retrieve connection report from agent: %v\n", err)
+				}
+			}
+
+			connDiags.Write(inv.Stdout)
 
 			n := 0
-			didP2p := false
 			start := time.Now()
+			results := make([]*ipnstate.PingResult, 0)
```

Comment:
Since this will now run indefinitely by default there's a large risk this will cause the process to eat up all available memory on the system.

For total, successful, min, max it's easy. For average you just need to also store a sum. For stddev it's a lot harder but you can use Welford's algorithm to calculate it without needing to store every value in memory.

API-wise I would add a method like `(s *pingSummary) addResult(result *ipnstate.PingResult)` which would do all the necessary updates after every ping attempt

Reply:
Very good point!!